File: neetcode-roadmap/stack/ValidParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];

  const dictionary = {
    '(': ')',
    '{': '}',
    '[': ']'
  }

  for (let i = 0; i < s.length; i++) {
    const char = s[i]

    if (char in dictionary) {
      stack.push(dictionary[char])
    } else {

      const top = stack.pop();
      if (top !== char) {
        return false;
      }

    }
  }

  return stack.length === 0
};

File: neetcode-roadmap/stack/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {"(": ")", "{": "}", "[": "]"}
        for i in s: 
            if i in dic: 
                stack.append(dic[i])
            else: 
                if not stack: 
                    return False
                top= stack.pop()
                if i != top: 
                    return False
    
        return len(stack) == 0 
